File: src/main/java/pl/training/workshop/m8/s08_compilergate/CompilerGate.java
package pl.training.workshop.m8.s08_compilergate;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;
import java.util.stream.Stream;

import javax.tools.Diagnostic;
import javax.tools.DiagnosticCollector;
import javax.tools.JavaCompiler;
import javax.tools.JavaFileObject;
import javax.tools.StandardJavaFileManager;
import javax.tools.ToolProvider;

/**
 * Bramka kompilatora: kompiluje wszystkie pliki katalogu przez javax.tools z opcjami
 * {@code --release 25 -Xlint:all -Werror} i zwraca werdykt oraz strukturalną diagnostykę
 * (kategoria lint, plik, linia) zamiast tekstu z konsoli.
 * <p>
 * Uwaga: z -Werror javac przerywa po fazie, w której padło pierwsze ostrzeżenie (np. fallthrough
 * z analizy przepływu już się nie pojawi). Dlatego pełną listę ostrzeżeń zbieramy osobnym
 * przebiegiem bez -Werror, a werdykt bierzemy z przebiegu z -Werror.
 */
public final class CompilerGate {
    public static final List<String> OPTIONS = List.of("--release", "25", "-proc:none", "-Xlint:all");

    public record Warning(String category, String file, long line) {
        @Override
        public String toString() {
            return "[" + category + "] " + file + ":" + line;
        }
    }

    public record Result(boolean passed, List<Warning> warnings) {
        public List<String> categories() {
            return warnings.stream().map(Warning::category).distinct().sorted().toList();
        }
    }

    private CompilerGate() {
    }

    public static Result check(Path sourceDir) {
        List<Path> sources = javaFiles(sourceDir);
        DiagnosticCollector<JavaFileObject> all = new DiagnosticCollector<>();
        compile(sources, List.of(), all);
        boolean passed = compile(sources, List.of("-Werror"), new DiagnosticCollector<>());
        return new Result(passed, warnings(all));
    }

    private static boolean compile(List<Path> sources, List<String> extraOptions,
            DiagnosticCollector<JavaFileObject> diagnostics) {
        JavaCompiler compiler = ToolProvider.getSystemJavaCompiler();
        Path output = createTempDir();
        try (StandardJavaFileManager files =
                compiler.getStandardFileManager(diagnostics, Locale.ROOT, StandardCharsets.UTF_8)) {
            List<String> options = new ArrayList<>(OPTIONS);
            options.addAll(extraOptions);
            options.addAll(List.of("-d", output.toString()));
            return compiler.getTask(null, files, diagnostics, options, null,
                    files.getJavaFileObjectsFromPaths(sources)).call();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        } finally {
            delete(output);
        }
    }

    private static List<Warning> warnings(DiagnosticCollector<JavaFileObject> diagnostics) {
        List<Warning> warnings = new ArrayList<>();
        for (Diagnostic<? extends JavaFileObject> d : diagnostics.getDiagnostics()) {
            boolean warning = d.getKind() == Diagnostic.Kind.WARNING
                    || d.getKind() == Diagnostic.Kind.MANDATORY_WARNING;
            if (warning && d.getSource() != null) {
                String category = lintCategory(d.getCode());
                String file = Path.of(d.getSource().toUri()).getFileName().toString();
                warnings.add(new Warning(category, file, d.getLineNumber()));
            }
        }
        return warnings;
    }

    /** Klucz komunikatu javac (np. compiler.warn.raw.class.use) na nazwę kategorii -Xlint. */
    private static String lintCategory(String code) {
        if (code.contains("deprecated.for.removal")) {
            return "removal";
        } else if (code.contains("deprecated")) {
            return "deprecation";
        } else if (code.contains("raw.class")) {
            return "rawtypes";
        } else if (code.contains("unchecked") || code.equals("compiler.warn.prob.found.req")) {
            return "unchecked";
        } else if (code.contains("fall-through")) {
            return "fallthrough";
        }
        return code;
    }

    private static List<Path> javaFiles(Path dir) {
        try (Stream<Path> files = Files.list(dir)) {
            return files.filter(f -> f.toString().endsWith(".java")).sorted().toList();
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private static Path createTempDir() {
        try {
            return Files.createTempDirectory("compiler-gate");
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

    private static void delete(Path dir) {
        try (Stream<Path> paths = Files.walk(dir)) {
            for (Path path : paths.sorted(Comparator.reverseOrder()).toList()) {
                Files.deleteIfExists(path);
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }
}

File: typescript/src/workshop/m8/s08_compilergate/CompilerGate.ts
import { readdirSync } from 'node:fs';
import path from 'node:path';

import ts from 'typescript-api';

/** Ostrzeżenie bramki: kategoria (jak -Xlint w javac), plik i linia. */
export class Warning {
  constructor(readonly category: string, readonly file: string, readonly line: number) {}

  toString(): string {
    return `[${this.category}] ${this.file}:${this.line}`;
  }
}

/** Werdykt bramki i pełna lista ostrzeżeń. */
export class Result {
  constructor(readonly passed: boolean, readonly warnings: readonly Warning[]) {}

  categories(): string[] {
    return [...new Set(this.warnings.map((warning) => warning.category))].sort();
  }
}

/**
 * Bramka kompilatora: kompiluje w pamięci wszystkie pliki katalogu przez API kompilatora
 * TypeScript (pakiet `typescript-api` = TypeScript 5.9) z opcjami ostrzejszymi niż build
 * projektu i zwraca werdykt oraz strukturalną diagnostykę (kategoria, plik, linia) zamiast
 * tekstu z konsoli. Każde ostrzeżenie jest traktowane jak błąd (odpowiednik -Werror).
 *
 * Kategorie (odpowiedniki -Xlint:all):
 * - rawtypes - typ generyczny utworzony bez argumentów typu, które TS uzupełnił jako any (new Map()),
 * - unchecked - niesprawdzona asercja typu (as) na wartości typu any,
 * - deprecation - wywołanie sygnatury oznaczonej w JSDoc jako @deprecated,
 * - fallthrough - przelot między case w switch (noFallthroughCasesInSwitch),
 * - unused, implicitReturns, unreachable, indexSignature - pozostałe opcje bramki.
 *
 * Uwaga: bramka nie honoruje dyrektyw @ts-expect-error i @ts-ignore - wyciszenie w kodzie
 * nie przepuszcza ostrzeżenia przez bramkę (build projektu je honoruje).
 */
export class CompilerGate {
  static readonly OPTIONS: Readonly<ts.CompilerOptions> = {
    target: ts.ScriptTarget.ES2023,
    module: ts.ModuleKind.NodeNext,
    moduleResolution: ts.ModuleResolutionKind.NodeNext,
    lib: ['lib.es2023.d.ts'],
    types: [],
    strict: true,
    noUncheckedIndexedAccess: true,
    exactOptionalPropertyTypes: true,
    noImplicitOverride: true,
    noFallthroughCasesInSwitch: true,
    noUnusedLocals: true,
    noUnusedParameters: true,
    noImplicitReturns: true,
    noPropertyAccessFromIndexSignature: true,
    allowUnreachableCode: false,
    allowUnusedLabels: false,
    skipLibCheck: true,
    noEmit: true,
  };

  private static readonly SUPPRESSION = /@ts-(?:expect-error|ignore|nocheck)/g;

  private constructor() {}

  static check(sourceDir: string): Result {
    const sources = CompilerGate.sourceFiles(sourceDir);
    const program = ts.createProgram({
      rootNames: sources,
      options: CompilerGate.OPTIONS,
      host: CompilerGate.hostWithoutSuppressions(),
    });
    const warnings: Warning[] = [];
    const checker = program.getTypeChecker();
    for (const fileName of sources) {
      const sourceFile = program.getSourceFile(fileName);
      if (sourceFile === undefined) {
        continue;
      }
      const diagnostics = [
        ...program.getSyntacticDiagnostics(sourceFile),
        ...program.getSemanticDiagnostics(sourceFile),
      ];
      for (const diagnostic of diagnostics) {
        warnings.push(CompilerGate.warning(CompilerGate.category(diagnostic.code), sourceFile, diagnostic.start ?? 0));
      }
      warnings.push(...CompilerGate.lint(sourceFile, checker));
    }
    warnings.sort((a, b) => a.file.localeCompare(b.file) || a.line - b.line || a.category.localeCompare(b.category));
    return new Result(warnings.length === 0, warnings);
  }

  /** Host czytający pliki z dysku, ale z "rozbrojonymi" dyrektywami wyciszającymi. */
  private static hostWithoutSuppressions(): ts.CompilerHost {
    const host = ts.createCompilerHost(CompilerGate.OPTIONS);
    const original = host.getSourceFile.bind(host);
    host.getSourceFile = (fileName, languageVersion, onError, shouldCreate) => {
      if (fileName.includes(`${path.sep}lib.`) || fileName.endsWith('.d.ts')) {
        return original(fileName, languageVersion, onError, shouldCreate);
      }
      const text = host.readFile(fileName);
      return text === undefined
        ? undefined
        : ts.createSourceFile(fileName, text.replace(CompilerGate.SUPPRESSION, 'ts-gate-ignored'), languageVersion);
    };
    return host;
  }

  /** Ostrzeżenia, których kompilator sam nie zgłasza: rawtypes, unchecked, deprecation. */
  private static lint(sourceFile: ts.SourceFile, checker: ts.TypeChecker): Warning[] {
    const warnings: Warning[] = [];
    const visit = (node: ts.Node): void => {
      if (ts.isNewExpression(node) && node.typeArguments === undefined && CompilerGate.hasAnyTypeArgument(node, checker)) {
        warnings.push(CompilerGate.warning('rawtypes', sourceFile, node.getStart()));
      }
      if (ts.isAsExpression(node) && (checker.getTypeAtLocation(node.expression).flags & ts.TypeFlags.Any) !== 0) {
        warnings.push(CompilerGate.warning('unchecked', sourceFile, node.getStart()));
      }
      if (ts.isCallExpression(node) || ts.isNewExpression(node)) {
        const signature = checker.getResolvedSignature(node);
        if (signature?.getJsDocTags().some((tag) => tag.name === 'deprecated')) {
          warnings.push(CompilerGate.warning('deprecation', sourceFile, node.getStart()));
        }
      }
      ts.forEachChild(node, visit);
    };
    visit(sourceFile);
    return warnings;
  }

  private static hasAnyTypeArgument(node: ts.NewExpression, checker: ts.TypeChecker): boolean {
    const type = checker.getTypeAtLocation(node);
    if ((type.flags & ts.TypeFlags.Object) === 0
      || ((type as ts.ObjectType).objectFlags & ts.ObjectFlags.Reference) === 0) {
      return false;
    }
    return checker.getTypeArguments(type as ts.TypeReference).some((argument) => (argument.flags & ts.TypeFlags.Any) !== 0);
  }

  /** Kod diagnostyki TypeScript na nazwę kategorii bramki. */
  private static category(code: number): string {
    switch (code) {
      case 7029: return 'fallthrough';
      case 6133: case 6138: case 6192: case 6196: case 6198: return 'unused';
      case 7030: return 'implicitReturns';
      case 7027: return 'unreachable';
      case 7028: return 'unusedLabel';
      case 4111: return 'indexSignature';
      default: return `TS${code}`;
    }
  }

  private static warning(category: string, sourceFile: ts.SourceFile, position: number): Warning {
    const line = sourceFile.getLineAndCharacterOfPosition(position).line + 1;
    return new Warning(category, path.basename(sourceFile.fileName), line);
  }

  private static sourceFiles(dir: string): string[] {
    return readdirSync(dir)
      .filter((name) => name.endsWith('.ts'))
      .sort()
      .map((name) => path.join(dir, name));
  }
}
